### PR TITLE
LC 2339 add domains to orgs

### PR DESCRIPTION
**Pull request by jaylco**
- View domains for organisation template
- Add domains to organisation functionality
- Introduce controller and route classes for greater reuse/testability
- Extend caching classes
- New validation middleware

### CODE REVIEW

File: src/csrs/model/organisationalUnit.ts [hunk 1]
```diff
@@ -59,4 +64,17 @@ export class OrganisationalUnit {
 		return (this.abbreviation && this.abbreviation !== '') ? `${this.name} (${this.abbreviation})` : this.name
 	}
 
+	insertAndSortDomain(domain: Domain) {
+		if (this.domains.find(d => d.domain === domain.domain) === undefined) {
```
Comment (SavvasCO) [inline src/csrs/model/organisationalUnit.ts:68]: You can use `.includes()` to see if an element exists in an array

File: src/csrs/model/organisationalUnit.ts [hunk 2]
```diff
@@ -59,4 +64,17 @@ export class OrganisationalUnit {
...
+			this.domains.push(domain)
+			this.domains.sort((a, b) => {
+				return (a.domain < b.domain) ? -1 : 1
+			})
+		}
+	}
+
+	doesDomainExist(domain: string) {
+		return Array.from(this.domains).find(d => d.domain === domain) !== undefined
```
Comment (SavvasCO) [inline src/csrs/model/organisationalUnit.ts:77]: You can use `includes()` here as well

**Review by SavvasCO — COMMENTED** [review]
LGTM. Just a suggestion.

File: src/csrs/model/organisationalUnit.ts (hunk 1 above)
Comment (jaylco) [reply src/csrs/model/organisationalUnit.ts:68]: `includes` compares stuff using ===, so even if the two domain objects being compared here are the same, they would actually be different as they're technically two different objects (one stored in the `this.domains` array and one the incoming object from the method).

I will however use `this.doesDomainExist` for this instead, which is doing the same thing and an oversight from me!

**Review by SavvasCO — APPROVED** [review]

**Pull request merged**